procedures: rewrite User Guide: Using a Git provider access token (#2509)
codenvy/che-docs: modules/end-user-guide/pages/using-a-git-provider-access-token.adoc

:_content-type: PROCEDURE
:description: Using a Git-provider access token
:keywords: Git, credentials, access-token
:navtitle: Using a Git-provider access token
:page-aliases: using-a-Git-credentials-store.adoc, using-git-credentials.adoc, 

[id="using-a-git-provider-access-token"]
= Using a Git-provider access token

> OAuth for GitHub, GitLab, or Bitbucket needs to be xref:administration-guide:oauth-for-github-gitlab-or-bitbucket.adoc[configured by the administrator] of your organization's {prod-short} instance. If it was not possible for your administrator to configure it for {prod-short} users, the workaround is for you to apply your personal access token as a {kubernetes} Secret.

Reviewer: ```suggestion
OAuth for GitHub, GitLab, or Bitbucket needs to be xref:administration-guide:oauth-for-github-gitlab-or-bitbucket.adoc[configured by the administrator] of your organization's {prod-short} instance. If your administrator could not configure it for {prod-short} users, the workaround is for you to apply your personal access token as a {kubernetes} Secret.
```

Mounting your access token as a Secret enables the {prod-short} Server to access the remote repository that is cloned during workspace creation, including access to the repository's `/.che` and `/.vscode` folders.

Apply the Secret in your user {orch-namespace} of the {orch-name} cluster of your organization's {prod-short} instance.

> After you have applied the Secret, you can create new workspaces from a private GitHub, GitLab, or Bitbucket-server repository.

Reviewer: ```suggestion
After applying the Secret, you can create new workspaces from a private GitHub, GitLab, or Bitbucket-server repository.
```

.Procedure

. Generate your access token on your Git provider's website.

. Encode your access token to Base64.
+
[TIP]
====
If you have the link:https://www.gnu.org/software/coreutils/base64[base64] command-line tools installed in the operating system, you can use the command line:

`$ echo -n '__<your_access_token_string>__' | base64`

====

. Visit `pass:c,a,q[{prod-url}]/api/user` in the web browser and copy the `id` value from the response. This is your {prod-short} user ID.

. Get your Git provider user ID by following the Git provider's API documentation:
+
* GitHub: link:https://docs.github.com/en/rest/users/users#get-a-user[Get a user]. See the `id` value in the response.
* GitLab: link:https://docs.gitlab.com/ee/api/users.html#for-normal-users[List users: For normal users], use the `username` filter: `/users?username=:username`. See the `id` value in the response.
* Bitbucket Server: link:https://developer.atlassian.com/server/bitbucket/rest/v802/api-group-api/#api-api-latest-users-userslug-get[Get users]. See the `id` value in the response.

. Prepare a new {orch-name} Secret.
+
[source,yaml,subs="+quotes,+attributes,+macros"]
----
kind: Secret
apiVersion: v1
metadata:

>   name: personal-access-token-__<your_chosen_name_for_this_token>__

Author: ```suggestion
  name: personal-access-token-__<your_choice_of_name_for_this_token>__
```

  labels:
    app.kubernetes.io/component: scm-personal-access-token
    app.kubernetes.io/part-of: che.eclipse.org
  annotations:

>     che.eclipse.org/che-userid: _<{prod-id-short}_user_ID>_# <1>

Author: ```suggestion
    che.eclipse.org/che-userid: __<{prod-id-short}_user_id>__# <1>
```

    che.eclipse.org/scm-personal-access-token-name: _<git_provider_name>_# <2>

>     che.eclipse.org/scm-url: _<Git_provider_endpoint># <3>

Author: ```suggestion
    che.eclipse.org/scm-url: __<git_provider_endpoint>__# <3>
```

>     che.eclipse.org/scm-userid: '_<Git_provider_user_ID>_'# <4>

Author: ```suggestion
    che.eclipse.org/scm-userid: '__<git_provider_user_id>__'# <4>
```

>     che.eclipse.org/scm-username: _<Git_provider_username>_

Author: ```suggestion
    che.eclipse.org/scm-username: __<git_provider_username>__
```

data:
  token: __<Base64_encoded_access_token>__
type: Opaque
----
+
<1> Your {prod-short} user ID.
<2> The Git provider name: `github` or `gitlab` or `bitbucket-server`.
<3> The Git provider URL.
<4> Your Git provider user ID.

. Visit `pass:c,a,q[{prod-url}]/api/kubernetes/namespace` to get your {prod-short} user namespace as `name`.

. Switch to your {prod-short} user namespace in the cluster.
+
[TIP]
====
On OpenShift:

* The `oc` command-line tool can return the namespace you are currently on in the cluster, which you can use to check your current namespace:
+
`$ oc project`

* You can switch to your {prod-short} user namespace on a command line if needed:
+
`$ oc project __<your_user_namespace>__`

====

. Apply the Secret.
+
[TIP]
====
On OpenShift, you can use the `oc` command-line tool:
[source,subs="+quotes,+attributes"]
----
$ oc apply -f - <<EOF
__<Secret_prepared_in_step_5>__
EOF
----
====

.Verification

> . xref:starting-a-new-workspace-with-a-clone-of-a-git-repository.adoc[Start a new workspace by using the URL of a remote Git repository] that is hosted by that Git provider.

Reviewer: ```suggestion
. xref:starting-a-new-workspace-with-a-clone-of-a-git-repository.adoc[Start a new workspace by using the URL of a remote Git repository] that the Git provider hosts.
```
